io/jvm/src/main/scala/scalaz/effect/RTS.scala

```diff
@@ -463,25 +455,16 @@ private object RTS {
                       }
                     } else {
                       // Error caught:
-                      val value = -\/(error)
-
-                      if (finalizer eq null) {
-                        // No finalizer to run:
-                        curIo = nextInstr[E](value, stack)
+                      curIo = nextInstr[E](error, stack)
```

Comment:
@ktonga 

Ok, after thinking about this, it's not quite correct: the reason is that handling of the error must occur _after_ finalization, not before. In this case, the `IO` representing the handling is created _before_ finalization, and may therefore reflect the wrong state.

So don't do this assignment. Delete this line entirely. Instead, we leave the continuation on the stack.

Then below, there are two cases:

1. There is no finalizer. In this case, we do: `curIo = IO.now(error)`. The loop will continue and the (error) value will trigger looking at the stack to find the next continuation, which will be the error handler.
2. There is a finalizer. In this case, we do: `curIo = ensuringUninterruptibleExit(finalization[E]) *> IO.now(error)`. This will perform finalization, then produce the error, then finally continue with the error handler on the stack, preserving linearity.

I cannot guarantee this will fix the test failure, there may be other problems, but this is definitely one.

Reply:
Thank you @jdegoes, I understand it better now. What you propose makes total sense. I pushed the change, didn't fix the tests tho.

Reply:
You can see tests results in the build checks details, I re-enabled them in my last push.
I'll keep looking tonight.

Reply:
All of the errors are in `bracket`, which may imply that it had some subtle dependency on the old error handling.

Reply:
Some more context. `testBracketRethrownCaughtErrorInAcquisition` is one of the tests failing.

Reply:
Yeah, instead of throwing the expected error it produces `()` as an `A` which I think is product of handling the error with the `finalizer` instead of using the `errorHandler` provided by `attempt`.

Shouldn't we turn [this](https://github.com/scalaz/scalaz-ioz/pull/5/files#diff-065b51e55bd0e079d39cdbe2d63585eeR467) around?

```
curIo = ensuringUninterruptibleExit(finalization[E]) *> curIo
```

Reply:
I'm not with my laptop now, but i'll keep debugging later when I get home.

Reply:
BTW how do you feel about adding a shit load of tracing using something macro-based only enabled at development by an env var?
Debugging a stack full of functions (which btw don't have the most useful toString) can be really hard to follow when hunting bugs.

Reply:
Change the code as follows and it should fix the problem (or at least get us closer):

```scala
                      curIo = nextInstr[E](error, stack)

                      if (finalizer ne null) {
                        // Must run finalizer first:
                        val finalization = dispatchErrors(finalizer)

                        // Do not interrupt finalization:
                        this.noInterrupt += 1

                        curIo = ensuringUninterruptibleExit(finalization[E]) *> curIo
                      }
```

Reply:
Didn't work.

Now we go one step further but then it fails when we invoke the `flatMap` from `absolve` which expects an `Either` but we pass a `()` produced by the finalizer. But that seem unrelated to the new use of the stack for the handler, b/c we are popping it right after pushing it in `catchError`.

Will keep trying tonight with a bit more of energy since it is friday here :smiley:

Reply:
@jdegoes  Finally I found the F#$%ing problem, it was due to A Series of Unfortunate Events:

`... *> curIo` was receiving `curIo` by-name but it's a `var`! so at the moment of evaluation it was a totally different effect :sob:. So now I'm capturing the desired effect in a `val` and bye bye bug.

Tomorrow I will go through all the docs and comments and will update them according to the changes, that would let the PR in good shape to be merged I think.

Reply:
@ktonga Brilliant work! Combination of two things, the `var` and by-name.